### PR TITLE
Fix the issue regarding custom theme

### PR DESCRIPTION
**Pull request by nilasini**
## Automation tests
 - Unit tests 
   > Code coverage information
 - Integration tests
   > Details about the test cases and coverage

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes/no
 - Ran FindSecurityBugs plugin and verified report? yes/no
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes/no

### CODE REVIEW

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/ThemeManagerHostObject.java [hunk 1]
```diff
@@ -380,6 +382,21 @@ private static void deployCustomTheme(FileHostObject themeFile, String tenantDom
         }
     }
 
+    private static String validateFilename(String filename, String intendedDir)
+            throws java.io.IOException {
+
+        File f = new File(filename);
+        String canonicalPath = f.getCanonicalPath();
+
+        File iD = new File(intendedDir);
```
Comment (dulanjal) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/ThemeManagerHostObject.java]: better to have more meaningful names for the 'f' and 'iD' variables.

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/ThemeManagerHostObject.java [hunk 2]
```diff
@@ -380,6 +382,21 @@ private static void deployCustomTheme(FileHostObject themeFile, String tenantDom
...
+        String canonicalID = iD.getCanonicalPath();
```
Comment (dulanjal) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/ThemeManagerHostObject.java]: better to have more meaningful names for the 'f' and 'iD' variables. It is hard to understand what those names mean without checking the variable declaration. Also, require corresponding meaningful names for 'canonicalPath' and 'canonicalID'

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/ThemeManagerHostObject.java [hunk 3]
```diff
@@ -338,7 +338,9 @@ private static void deployCustomTheme(FileHostObject themeFile, String tenantDom
 
             while (ze != null) {
                 String fileName = ze.getName();
-                Path newFilePath = themePath.resolve(fileName);
+                String intendedDir = themePath.toAbsolutePath().toString();
```
Comment (dulanjal) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/ThemeManagerHostObject.java]: Is it necessary to call 'toAbsolutePath'? Canonicalization in validateFilename() should automatically take the absolute path.

**Review by malithie — APPROVED** [review]

**Review by dulanjal — APPROVED** [review]

File: components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/ThemeManagerHostObject.java [hunk 4]
```diff
@@ -380,6 +382,21 @@ private static void deployCustomTheme(FileHostObject themeFile, String tenantDom
...
+        File file = new File(filename);
+        String systemDependentFilePath = file.getCanonicalPath();
+
+        File intendedDirectory = new File(intendedDir);
+        String systemDependentIntendedDirectoryPath = intendedDirectory.getCanonicalPath();
+
+        if (systemDependentFilePath.startsWith(systemDependentIntendedDirectoryPath)) {
+            return systemDependentFilePath;
+        } else {
+            throw new IllegalStateException("File is outside extraction target directory.");
```
Comment (dulanjal) [inline components/org.wso2.carbon.appmgt.hostobjects/src/main/java/org/wso2/carbon/appmgt/hostobjects/ThemeManagerHostObject.java]: Let's print the original name of the zip entry with the exception message

**Review by kasundharmadasa — APPROVED** [review]

**Pull request merged**